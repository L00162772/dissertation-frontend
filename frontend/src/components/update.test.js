import React from 'react';
import { shallow } from 'enzyme';
import {render, screen, fireEvent, cleanup} from '@testing-library/react';
import Update from './update';
import {BrowserRouter as Router} from 'react-router-dom';
import axios from 'axios'
jest.mock('axios');

afterEach(cleanup);
afterEach(() => {
    axios.put.mockClear();
  });

  function mockPutDataCall() {
    axios.put.mockResolvedValueOnce({
      data: {
        results: [
          {
            name: {
              first: "name 1"
            }
          },
          {
            name: {
              first: "name 2"
            }
          }
        ]
      }
    });
  }

describe('<Update /> with no props', () => {
  const container = shallow(<Router><Update /></Router>);

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });
  
  // See https://testing-library.com/docs/react-testing-library/cheatsheet
  //https://testing-library.com/docs/example-input-event/
  it('should have firstName field', () => {
    render(<Router><Update /></Router>);
    const input = screen.getByTestId('firstName');
    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('placeholder',  'First Name')
    expect(input).toHaveAttribute('type',  'text')
  });

  it('should accept changes in firstName field', () => {
    render(<Router><Update /></Router>);
    const input = screen.getByTestId('firstName');
    fireEvent.change(input, {target: {value: 'first name'}})
    expect(input.value).toBe('first name')
  });

  it('should have lastName field', () => {
    render(<Router><Update /></Router>);
    const input = screen.getByTestId('lastName');
    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('placeholder',  'Last Name')
    expect(input).toHaveAttribute('type',  'text')
  });

  it('should accept changes in lastName field', () => {
    render(<Router><Update /></Router>);
    const input = screen.getByTestId('lastName')
    fireEvent.change(input, {target: {value: 'last name'}})
    expect(input.value).toBe('last name')
  });

  it('should have terms field', () => {
    render(<Router><Update /></Router>);
    const input = screen.getByTestId('terms')
    expect(input).toBeInTheDocument()
    const checkboxLabel = input.children[1]
    expect(checkboxLabel).toHaveTextContent("I agree to the Terms and Conditions")
  });

  it('should accept changes to terms checkbox', () => {
    render(<Router><Update /></Router>);
    const input= screen.getByTestId('terms') 

    const checkboxInput = input.children[0]
    fireEvent.click(checkboxInput)
    expect(input).toHaveAttribute('class',  'ui checked checkbox')

    fireEvent.click(checkboxInput)
    expect(input).toHaveAttribute('class',  'ui checkbox')
  });

  it('should have submit button', () => {
    render(<Router><Update /></Router>);
    const button = screen.getByTestId('update')
    expect(button).toBeInTheDocument()
    expect(button).toHaveAttribute('type',  'submit')
  });

  it('should trigger submit button', () => {
    render(<Router><Update /></Router>);
    const button = screen.getByTestId('update')

    mockPutDataCall();

    fireEvent.click(button)
  });
});