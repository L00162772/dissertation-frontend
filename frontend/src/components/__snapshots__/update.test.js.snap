// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Update /> with no props should match the snapshot 1`] = `"<div><form class=\\"ui form create-form\\"><div class=\\"field\\"><label>First Name</label><input placeholder=\\"First Name\\" type=\\"text\\" data-testid=\\"firstName\\" value=\\"\\"/></div><div class=\\"field\\"><label>Last Name</label><input placeholder=\\"Last Name\\" type=\\"text\\" data-testid=\\"lastName\\" value=\\"\\"/></div><div class=\\"field\\"><div data-testid=\\"terms\\" class=\\"ui checkbox\\"><input class=\\"hidden\\" readonly=\\"\\" tabindex=\\"0\\" type=\\"checkbox\\"/><label>I agree to the Terms and Conditions</label></div></div><button data-testid=\\"update\\" class=\\"ui button\\" type=\\"submit\\">Update</button></form></div>"`;
